### PR TITLE
Authentication and Page Order Modules

### PR DESCRIPTION
**Pull request by ch0nx**
##### SUMMARY

In order to fill some gaps in IIS website provisioning, specifically around Citrix Storefront deployments, our team had developed a few custom modules to configure authentication types and the website page order. Instead of keeping them internal we thought we should share!

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
microsoft.iis.page_order
microsoft.iis.authentication

##### ADDITIONAL INFORMATION
microsoft.iis.authentication module:
Intention:
Configures IIS authentication settings, supporting multiple authentication types (e.g., WindowsAuthentication, AnonymousAuthentication). It allows enabling/disabling authentication, setting providers, and configuring advanced options like usekernelmode and tokenchecking for Windows Authentication.

Summary:
- Uses PowerShell’s WebAdministration module to query and set IIS authentication configuration.
  - Since WIndowsAuthentication has different options available, it has a separate set of logic that is executed when configuring this type of authentication.
- Accepts parameters for ps_path, location, auth_type, enabled, providers, usekernelmode, and tokenchecking.
- Validates required parameters, especially for Windows Authentication, where additional options are mandatory.
- Implements idempotency: checks current state before applying changes, and only updates if necessary.
- Supports Ansible check mode for dry-run capability.
- Returns detailed diff and status messages for reporting and troubleshooting.
- ansible-test support for basic and edge cases
  - Loops through a list of auth configs for basic testing, including setting it to the opposite, validating check mode changes, then setting it to the desired state.

microsoft.iis.page_order module:
Intention:
Manages the default document order in IIS by configuring the order of default pages (e.g., default.aspx, Default.htm) for a given site or path.

Key Features:

- Accepts parameters for ps_path or website_path (mutually exclusive), page_order (comma-separated), filter, and collection_name.
- Retrieves current page order, compares with the desired state, and updates only if there is a mismatch.
- Removes existing page order before setting a new one to ensure correct ordering.
- Uses PowerShell’s WebAdministration module for all operations.
- Supports Ansible check mode and provides before/after diffs and status messages.
- ansible-test support for module functionality

ansible-test results in the attached txt files, targeting a server 2022DC machine:
[authentication.txt](https://github.com/user-attachments/files/22052901/authentication.txt)
[page_order.txt](https://github.com/user-attachments/files/22052902/page_order.txt)

Extra notes:
- This was my first foray into creating tests for ansible-test specifically, but it does appear to work as intended and I tried to keep it in line with the functionality currently in place for other modules.
  - IIS setup/teardown for the tests should be in line with the other modules
  - The documentation for running the tests in this repo is incorrect, since using just 'integration' does not target windows hosts, it has to be windows-integration
- I thought it was odd that the other doc files were in yml format when the [docs](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general_windows.html#windows-new-module-development) say that it should be in .py, but it looks like I got them converted to something that the builder accepts now. I got the PSLint errors resolved by matching the style to the current, instead of the formatting I had previously. Broke up the longer lines as well.
- Apparently if you get false positives for no-log-needed, you have to explicitly set no_log to false in the argument spec.

### CODE REVIEW

**Comment by jborean93** [comment]
Thanks for the PR, I've got it marked to look at tomorrow!

**Review by jborean93 — CHANGES_REQUESTED** [review]
I've gone over an initial pass and most of my comments from `authentication` apply to `page_order`. I think the documentation might need some expansion on some of the options to really clarify what they are for and what they can be set to. The docs should also be updated to use the Ansible flavoured markup values as documented under https://docs.ansible.com/ansible/latest/dev_guide/ansible_markup.html#ansible-markup.

A few of the options really seem to just be the underlying cmdlets mapped to an Ansible option. We want to expose a more human friendly interface so things like `website` can be used to find by name rather than something like `pa_path`.

For the testing I would like to see something a bit more expanded to cover things like checking that in check mode things reported a change but no change actually occurred.

I know some of this stuff might sound pedantic and onerous but this is a certified collection from Red Hat so the bar for acceptance is quite high. If you need any help around how we do things/testing guidelines etc please let me know and I'll try and help as best as I can. My knowledge of IIS is a bit limited here so some of my questions may sound a bit dumb as I don't really have the management context behind this product.

File: plugins/modules/authentication.ps1 [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#AnsibleRequires -CSharpUtil Ansible.Basic
+#AnsibleRequires -PowerShell Ansible.ModuleUtils.AddType
+
+# Note: If troubleshooting, use $DebugPreference = 'Continue' and Start-Transcript <filepath>
+# in order to get debug output to a readable location. Ansible does not store debug stream output.
+$ErrorActionPreference = 'Stop'
+$spec = @{
+    options = @{
+        ps_path = @{ default = 'IIS:\'; required = $false; type = 'str' }
```
[the remark this answers is not in the document]
Comment (ch0nx) [reply plugins/modules/authentication.ps1:12]: The auth configuration in IIS gets pretty picky with the paths, even to the point that specifying the auth methods on a sub path without modifying the root IIS:\ level can result in errors. In this case I had opted to leave it named ps_path since it was what was being passed as the PSPath parameter to the Set-WebConfiguration and Set-WebConfigurationProperty cmdlets. I'm not married to it though if you think it should be restricted or changed in some other way

File: plugins/modules/authentication.ps1 [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+$psPath = $module.Params.ps_path
+$location = $module.Params.location
+$authType = $module.Params.auth_type
+$enabled = $module.Params.enabled
+$Providers = $module.Params.providers
+$UseKernelMode = $module.Params.usekernelmode
+$tokenChecking = $module.Params.tokenchecking
+$module.Result.diff = @{
+    before = ''
+    after = ''
```
[the remark this answers is not in the document]
Comment (ch0nx) [reply plugins/modules/authentication.ps1:45]: When I was looking at other custom PS modules that support diff mode I was only seeing these values as strings, before I start plugging this in was there some other format that this should follow? I can definitely at least get the values to line up with the params instead of just showing what matched

File: plugins/modules/page_order.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      required: true
+    filter:
+      description:
+        - "Filter to be passed to Get-WebConfigurationProperty. Example: 'system.webServer/defaultDocument/files'."
+      type: str
+      required: false
+      default: 'system.webServer/defaultDocument/files'
+    collection_name:
+      description:
+        - "The 'collection' name that is being targeted, this gets appended to the 'filter' parameter when getting the current value with Get-WebConfigurationProperty. Example: 'add'."
```
[the remark this answers is not in the document]
Comment (ch0nx) [reply plugins/modules/page_order.yml:36]: Good call - I will change it to just use the filter and restructure accordingly. It was just there as a "in case something needs it" kind of thing.

File: plugins/modules/authentication.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        - "Value to be used for the 'TokenChecking' option for WindowsAuthentication extended protection."
+      type: str
+      required: false
+  extends_documentation_fragment:
+    - ansible.builtin.action_common_attributes
+  attributes:
+    check_mode:
+      support: full
+    diff_mode:
+      support: none
```
[the remark this answers is not in the document]
Comment (ch0nx) [reply plugins/modules/authentication.yml]: Fixed in both

File: plugins/modules/authentication.yml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      - Dictionary showing before and after values for authentication settings.
+    returned: always
+    type: dict
+    sample:
+      before:
+        enabled: false
+        providers: "Negotiate"
+      after:
+        enabled: true
+        providers: "Negotiate:Kerberos"
```
[the remark this answers is not in the document]
Comment (ch0nx) [reply plugins/modules/authentication.yml]: Removed in both

File: plugins/modules/authentication.ps1 [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+
+        [Parameter(Mandatory = $false, ValueFromPipelineByPropertyName = $true)]
+        [bool]
+        $UseKernelMode,
+
+        [Parameter(Mandatory = $false, ValueFromPipelineByPropertyName = $true)]
+        [string]
+        $TokenChecking,
+
+        [Parameter(Mandatory = $false, ValueFromPipelineByPropertyName = $true)]
```
[the remark this answers is not in the document]
Comment (ch0nx) [reply plugins/modules/authentication.ps1]: Removed - leftover from when this was a script

File: plugins/modules/authentication.ps1 [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+# in order to get debug output to a readable location. Ansible does not store debug stream output.
+$ErrorActionPreference = 'Stop'
+$spec = @{
+    options = @{
+        ps_path = @{ default = 'IIS:\'; required = $false; type = 'str' }
+        location = @{ required = $false; type = 'str' }
+        auth_type = @{ required = $true; type = 'str' }
+        enabled = @{ required = $true; type = 'bool' }
+        providers = @{ required = $false; type = 'str' }
+        usekernelmode = @{ required = $false; type = 'bool' }
```
[the remark this answers is not in the document]
Comment (ch0nx) [reply plugins/modules/authentication.ps1]: snakified

File: plugins/modules/authentication.ps1 [hunk 8]
```diff
@@ -0,0 +1,357 @@
+#!powershell
+# GNU General Public License v3.0+
+# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#AnsibleRequires -CSharpUtil Ansible.Basic
+#AnsibleRequires -PowerShell Ansible.ModuleUtils.AddType
+
+# Note: If troubleshooting, use $DebugPreference = 'Continue' and Start-Transcript <filepath>
+# in order to get debug output to a readable location. Ansible does not store debug stream output.
+$ErrorActionPreference = 'Stop'
```
[the remark this answers is not in the document]
Comment (ch0nx) [reply plugins/modules/authentication.ps1]: Removed

File: plugins/modules/authentication.ps1 [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+$authType = $module.Params.auth_type
+$enabled = $module.Params.enabled
+$Providers = $module.Params.providers
+$UseKernelMode = $module.Params.usekernelmode
+$tokenChecking = $module.Params.tokenchecking
+$module.Result.diff = @{
+    before = ''
+    after = ''
+}
+$module.Result.msg = ''
```
[the remark this answers is not in the document]
Comment (jborean93) [reply plugins/modules/authentication.ps1:45]: It's not something we typically encourage anymore as people start writing conditionals on `result.msg` meaning we can no longer change that message without potentially breaking people's playbooks. Unfortunately right now it's really just the `diff` output that we rely on this.

File: plugins/modules/authentication.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (jborean93) [reply plugins/modules/authentication.ps1:43]: It's unfortunately a bit over the place because not all modules implement diff and the ones that do have done it in a few different ways. An example of the diff before/after being a structured values can be found in [win_group](https://github.com/ansible-collections/ansible.windows/blob/9cb32d0c26b7fb99341cb6aef74691bcef007901/plugins/modules/win_group.ps1).

This is where the diff is first defined https://github.com/ansible-collections/ansible.windows/blob/9cb32d0c26b7fb99341cb6aef74691bcef007901/plugins/modules/win_group.ps1#L55-L56. We keep the defaults as `$null` so that if there was no existing resource/configuration before no action is needed and if we are removing the resource/configuration there is also no action needed.

We explicitly set the `before` value if the resource existed/had some configuration at https://github.com/ansible-collections/ansible.windows/blob/9cb32d0c26b7fb99341cb6aef74691bcef007901/plugins/modules/win_group.ps1#L150-L158. We then add the `after` here https://github.com/ansible-collections/ansible.windows/blob/9cb32d0c26b7fb99341cb6aef74691bcef007901/plugins/modules/win_group.ps1#L236-L242.

It's a long video but I've also got a video going over this in more detail at https://youtu.be/t28guLbd56U?t=3865&si=Wd7zDvITiJpM8yKs.

File: plugins/modules/authentication.ps1 [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+        ps_path = @{ default = 'IIS:\'; required = $false; type = 'str' }
+        location = @{ required = $false; type = 'str' }
+        auth_type = @{ required = $true; type = 'str' }
+        enabled = @{ required = $true; type = 'bool' }
+        providers = @{ required = $false; type = 'str' }
+        usekernelmode = @{ required = $false; type = 'bool' }
+        tokenchecking = @{ required = $false; type = 'str'; no_log = $false }
+    }
+    required_if = @(
+        , @('auth_type', 'WindowsAuthentication', @('providers', 'usekernelmode', 'tokenchecking'))
```
[the remark this answers is not in the document]
Comment (jborean93) [reply plugins/modules/authentication.ps1]: I've found it best to keep them option so there is a 3rd state:

+ None/null/unset/omit means you want to preserve the existing value
+ false means you want to disable it
+ true means you want to enable it

The first one is important as it means someone could use a module to set this configuration without knowing all the existing values or having to retrieve them in a previous step if they wanted to keep it the same but set other options.

**Comment by ch0nx** [comment]
My apologies, I completely spaced this module/PR. Updated based off of the current module contents and made the following changes:
- Converted to just use site targeting instead of ps path + location, with the module performing the interpretation behind the scenes. Originally this was just functioning as more of a wrapper for the webadministration functions and passing the params in directly, but this should be more user friendly.
- Changed diff based output based on feedback and removed `msg` output
- Changed to the three-state configuration that was requested in place of the parameter requirements, where it preserves existing settings or uses default values if not specified.
  - It will warn, but not fail, if WindowsAuthentication parameters are passed with non-WindowsAuthentication authentication methods
- Providers parameter changed to accept a list instead of a comma separated string
- Changed to have page_order accept a list, instead of a comma separated string.
- Added optional "application" that gets concatenated with the site name, for nested app targeting (previously covered by ps_path)
  - Note: nested applications that share the physical directory of its parent could display idempotency issues if they have a different page order specified from their parent, since the web.config is shared between them, apparently. 
- Added a `Compare-IISAuth` function for comparing the before/after states since the comparisons were being made multiple times
- Added extra guards for things like site/application existence
- Added some more tests for evaluating the different expected states
- Modified negative tests to align more with what the other modules were doing

**Comment by ch0nx** [comment]
recheck

**Comment by jborean93** [comment]
> My apologies, I completely spaced this module/PR

No worries, plan is to do a review early next week. Thanks for following up on this.

**Review by jborean93 — APPROVED** [review]
Sorry for the delay, I've been a bit busy with a few other things and this took a bit longer to get back to.
